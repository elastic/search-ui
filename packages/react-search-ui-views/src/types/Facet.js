import PropTypes from "prop-types";
import FacetType from "./FacetType";
import FacetValue from "./FacetValue";

export default PropTypes.shape({
  data: PropTypes.arrayOf(FacetValue).isRequired,
  field: PropTypes.string.isRequired,
  type: FacetType.isRequired
});
