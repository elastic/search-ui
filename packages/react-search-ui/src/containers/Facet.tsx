import React from "react";
import { Component } from "react";
import {
  FacetContainerContext,
  FacetContainerProps,
  FacetViewProps,
  MultiCheckboxFacet
} from "@elastic/react-search-ui-views";
import { helpers } from "@elastic/search-ui";

import { accentFold } from "../helpers";
import { withSearch } from "..";

const { markSelectedFacetValuesFromFilters } = helpers;

type FacetContainerState = {
  searchTerm: string;
  more: number;
};

export class FacetContainer extends Component<
  FacetContainerProps,
  FacetContainerState
> {
  static defaultProps = {
    filterType: "all",
    isFilterable: false,
    show: 5
  };

  constructor(props) {
    super(props);
    this.state = {
      more: props.show,
      searchTerm: ""
    };
  }

  handleClickMore = (totalOptions) => {
    this.setState(({ more }) => {
      let visibleOptionsCount = more + 10;
      const showingAll = visibleOptionsCount >= totalOptions;
      if (showingAll) visibleOptionsCount = totalOptions;

      this.props.a11yNotify("moreFilters", { visibleOptionsCount, showingAll });

      return { more: visibleOptionsCount };
    });
  };

  handleFacetSearch = (searchTerm) => {
    this.setState({ searchTerm });
  };

  render() {
    const { more, searchTerm } = this.state;
    const {
      addFilter,
      className,
      facets,
      field,
      filterType,
      filters,
      label,
      removeFilter,
      setFilter,
      view,
      isFilterable,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      a11yNotify,
      ...rest
    } = this.props;
    const facetsForField = facets[field];

    if (!facetsForField) return null;

    // By using `[0]`, we are currently assuming only 1 facet per field. This will likely be enforced
    // in future version, so instead of an array, there will only be one facet allowed per field.
    const facet = facetsForField[0];

    let facetValues = markSelectedFacetValuesFromFilters(
      facet,
      filters,
      field,
      filterType
    ).data;

    const selectedValues = facetValues
      .filter((fv) => fv.selected)
      .map((fv) => fv.value);

    if (!facetValues.length && !selectedValues.length) return null;

    if (searchTerm.trim()) {
      facetValues = facetValues.filter((option) => {
        let valueToSearch;
        switch (typeof option.value) {
          case "string":
            valueToSearch = accentFold(option.value).toLowerCase();
            break;
          case "number":
            valueToSearch = option.value.toString();
            break;
          case "object":
            valueToSearch =
              typeof option?.value?.name === "string"
                ? accentFold(option.value.name).toLowerCase()
                : "";
            break;

          default:
            valueToSearch = "";
            break;
        }
        return valueToSearch.includes(accentFold(searchTerm).toLowerCase());
      });
    }

    const View: React.ComponentType<FacetViewProps> =
      view || MultiCheckboxFacet;

    const viewProps: FacetViewProps = {
      className,
      label: label,
      onMoreClick: this.handleClickMore.bind(this, facetValues.length),
      onRemove: (value) => {
        removeFilter(field, value, filterType);
      },
      onChange: (value) => {
        setFilter(field, value, filterType);
      },
      onSelect: (value) => {
        addFilter(field, value, filterType);
      },
      options: facetValues.slice(0, more),
      showMore: facetValues.length > more,
      values: selectedValues,
      showSearch: isFilterable,
      onSearch: (value) => {
        this.handleFacetSearch(value);
      },
      searchPlaceholder: `Filter ${label}`,
      ...rest
    };

    return <View {...viewProps} />;
  }
}

export default withSearch<FacetContainerProps, FacetContainerContext>(
  ({ filters, facets, addFilter, removeFilter, setFilter, a11yNotify }) => ({
    filters,
    facets,
    addFilter,
    removeFilter,
    setFilter,
    a11yNotify
  })
)(FacetContainer);
