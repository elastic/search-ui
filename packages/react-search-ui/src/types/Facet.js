import PropTypes from "prop-types";
import FacetType from "./FacetType";
import FacetValue from "./FacetValue";

export default PropTypes.shape({
  data: PropTypes.arrayOf(FacetValue).isRequired,
  // Name of the field this facet is associated with
  field: PropTypes.string.isRequired,
  type: FacetType.isRequired
});
