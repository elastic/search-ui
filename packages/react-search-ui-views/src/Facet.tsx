import {
  FacetValue,
  FieldValue,
  FilterType,
  SearchContextState
} from "@elastic/search-ui";
import { BaseContainerProps } from "./types";

export type FacetViewProps = {
  className?: string;
  label: string;
  onMoreClick: () => void;
  onRemove: (value: FieldValue) => void;
  onChange: (value: FieldValue) => void;
  onSelect: (value: FieldValue) => void;
  options: FacetValue[];
  showMore: boolean;
  values: any;
  showSearch: boolean;
  onSearch: (value: string) => void;
  searchPlaceholder: string;
};

export type FacetContainerContext = Pick<
  SearchContextState,
  | "addFilter"
  | "removeFilter"
  | "setFilter"
  | "a11yNotify"
  | "filters"
  | "facets"
>;

export type FacetContainerProps = BaseContainerProps & {
  filterType: FilterType;
  show?: number;
  view?: React.ComponentType<FacetViewProps>;
  isFilterable: boolean;
  field: string;
  label: string;
} & FacetContainerContext;
