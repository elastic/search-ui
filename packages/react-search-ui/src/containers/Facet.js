import PropTypes from "prop-types";
import { Component } from "react";
import { MultiCheckboxFacet } from "@elastic/react-search-ui-views";
import { helpers } from "@elastic/search-ui";

import { Facet, Filter, FilterType } from "../types";
import { accentFold } from "../helpers";
import { withSearch } from "..";

const { markSelectedFacetValuesFromFilters } = helpers;

export class FacetContainer extends Component {
  static propTypes = {
    // Props
    className: PropTypes.string,
    field: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    filterType: FilterType,
    show: PropTypes.number,
    view: PropTypes.func,
    isFilterable: PropTypes.bool,
    // State
    filters: PropTypes.arrayOf(Filter).isRequired,
    facets: PropTypes.objectOf(PropTypes.arrayOf(Facet)).isRequired,
    // Actions
    addFilter: PropTypes.func.isRequired,
    removeFilter: PropTypes.func.isRequired,
    setFilter: PropTypes.func.isRequired,
    a11yNotify: PropTypes.func.isRequired
  };

  static defaultProps = {
    filterType: "all",
    isFilterable: false
  };

  constructor({ show = 5 }) {
    super();
    this.state = {
      more: show,
      searchTerm: ""
    };
  }

  handleClickMore = totalOptions => {
    this.setState(({ more }) => {
      let visibleOptionsCount = more + 10;
      const showingAll = visibleOptionsCount >= totalOptions;
      if (showingAll) visibleOptionsCount = totalOptions;

      this.props.a11yNotify("moreFilters", { visibleOptionsCount, showingAll });

      return { more: visibleOptionsCount };
    });
  };

  handleFacetSearch = searchTerm => {
    this.setState({ searchTerm });
  };

  render() {
    const { more, searchTerm } = this.state;
    const {
      addFilter,
      className,
      facets,
      field,
      filterType,
      filters,
      label,
      removeFilter,
      setFilter,
      view,
      isFilterable,
      // eslint-disable-next-line no-unused-vars
      a11yNotify,
      ...rest
    } = this.props;
    const facetsForField = facets[field];

    if (!facetsForField) return null;

    // By using `[0]`, we are currently assuming only 1 facet per field. This will likely be enforced
    // in future version, so instead of an array, there will only be one facet allowed per field.
    const facet = facetsForField[0];

    let facetValues = markSelectedFacetValuesFromFilters(
      facet,
      filters,
      field,
      filterType
    ).data;

    const selectedValues = facetValues
      .filter(fv => fv.selected)
      .map(fv => fv.value);

    if (!facetValues.length && !selectedValues.length) return null;

    if (searchTerm.trim()) {
      facetValues = facetValues.filter(option =>
        accentFold(option.value)
          .toLowerCase()
          .includes(accentFold(searchTerm).toLowerCase())
      );
    }

    const View = view || MultiCheckboxFacet;

    return View({
      className,
      label: label,
      onMoreClick: this.handleClickMore.bind(this, facetValues.length),
      onRemove: value => {
        removeFilter(field, value, filterType);
      },
      onChange: value => {
        setFilter(field, value, filterType);
      },
      onSelect: value => {
        addFilter(field, value, filterType);
      },
      options: facetValues.slice(0, more),
      showMore: facetValues.length > more,
      values: selectedValues,
      showSearch: isFilterable,
      onSearch: value => {
        this.handleFacetSearch(value);
      },
      searchPlaceholder: `Filter ${field}`,
      ...rest
    });
  }
}

export default withSearch(
  ({ filters, facets, addFilter, removeFilter, setFilter, a11yNotify }) => ({
    filters,
    facets,
    addFilter,
    removeFilter,
    setFilter,
    a11yNotify
  })
)(FacetContainer);
