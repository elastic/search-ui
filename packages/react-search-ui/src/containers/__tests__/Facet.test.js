import React from "react";
import { FacetContainer } from "../Facet";
import { shallow } from "enzyme";

const View = () => <div />;

const params = {
  view: jest.fn().mockImplementation(props => <View {...props} />),
  field: "field1",
  label: "Field 1",
  filters: [
    {
      field: "field1",
      values: ["field1value1"],
      type: "all"
    }
  ],
  facets: {
    field1: [
      {
        field: "field1",
        data: [
          { count: 20, value: "field1value1" },
          { count: 10, value: "field1value2" }
        ],
        type: "value"
      }
    ],
    field2: [
      {
        field: "field2",
        data: [{ count: 10, value: "field2value1" }],
        type: "value"
      }
    ],
    field3: [
      {
        field: "field3",
        data: [],
        type: "value"
      }
    ]
  },
  addFilter: jest.fn(),
  removeFilter: jest.fn(),
  setFilter: jest.fn(),
  a11yNotify: jest.fn()
};

beforeEach(() => {
  jest.clearAllMocks();
});

it("renders correctly", () => {
  const wrapper = shallow(<FacetContainer {...params} />);
  expect(wrapper).toMatchSnapshot();
});

it("should not render a Facet if there are no facets available", () => {
  const wrapper = shallow(
    <FacetContainer
      {...{
        ...params,
        facets: {}
      }}
    />
  );

  expect(wrapper.html()).toEqual(null);
});

describe("values view prop", () => {
  function subject(filterType) {
    let viewProps;
    shallow(
      <FacetContainer
        {...params}
        field="field1"
        filterType={filterType}
        filters={[
          {
            field: "field2",
            values: ["field1value1"],
            type: "none"
          },
          {
            field: "field1",
            values: ["field1value1", "field1value2"],
            type: "all"
          },
          {
            field: "field1",
            values: ["field1value1"],
            type: "any"
          }
        ]}
        facets={{
          field1: [
            {
              field: "field1",
              data: [
                { count: 20, value: "field1value1" },
                { count: 10, value: "field1value2" }
              ],
              type: "value"
            }
          ]
        }}
        view={props => (viewProps = props)}
      />
    );
    return viewProps;
  }
  it("should correctly calculated the selected facet options", () => {
    const { values } = subject();
    expect(values).toEqual(["field1value1", "field1value2"]);
  });

  it("should correctly calculated the selected facet options for filterType", () => {
    const { values } = subject("any");
    expect(values).toEqual(["field1value1"]);
  });
});

describe("show more", () => {
  let wrapper;

  function subject(additionalProps = {}) {
    return shallow(
      <FacetContainer
        {...{
          ...params,
          field: "field1",
          facets: {
            field1: [
              {
                field: "field1",
                data: [
                  { count: 20, value: "field1value1" },
                  { count: 10, value: "field1value2" },
                  { count: 9, value: "field1value3" },
                  { count: 8, value: "field1value4" },
                  { count: 7, value: "field1value5" },
                  { count: 6, value: "field1value6" },
                  { count: 5, value: "field1value7" },
                  { count: 4, value: "field1value8" },
                  { count: 3, value: "field1value9" },
                  { count: 2, value: "field1value10" },
                  { count: 1, value: "field1value11" },
                  { count: 1, value: "field1value12" },
                  { count: 1, value: "field1value13" },
                  { count: 1, value: "field1value14" },
                  { count: 1, value: "field1value15" },
                  { count: 1, value: "field1value16" },
                  { count: 1, value: "field1value17" }
                ],
                type: "value"
              }
            ]
          },
          ...additionalProps
        }}
      />
    );
  }

  beforeAll(() => {
    wrapper = subject();
  });

  it("should limit options to 5 initially", () => {
    expect(wrapper.find(View).prop("options").length).toEqual(5);
  });

  it("should start at more than 5 initially if a show prop is passed", () => {
    const initialCount = 10;
    const newWrapper = subject({ show: initialCount });
    expect(newWrapper.find(View).prop("options").length).toEqual(initialCount);
  });

  it("should have a show more button", () => {
    expect(wrapper.find(View).prop("showMore")).toEqual(true);
  });

  describe("after a show more click", () => {
    beforeAll(() => {
      wrapper.find(View).prop("onMoreClick")();

      expect(params.a11yNotify).toHaveBeenCalledWith("moreFilters", {
        visibleOptionsCount: 15,
        showingAll: false
      });
    });

    it("should have 10 more options", () => {
      expect(wrapper.find(View).prop("options").length).toEqual(15);
    });

    it("should still have a show more button", () => {
      expect(wrapper.find(View).prop("showMore")).toEqual(true);
    });
  });

  describe("after more more show more click", () => {
    beforeAll(() => {
      wrapper.find(View).prop("onMoreClick")();

      expect(params.a11yNotify).toHaveBeenCalledWith("moreFilters", {
        visibleOptionsCount: 17,
        showingAll: true
      });
    });

    it("should be showing all options", () => {
      expect(wrapper.find(View).prop("options").length).toEqual(17);
    });

    it("should hide the show more button", () => {
      expect(wrapper.find(View).prop("showMore")).toEqual(false);
    });
  });
});

it("will add a filter when a facet value is selected with onSelect", () => {
  const wrapper = shallow(<FacetContainer {...params} filterType="any" />);

  wrapper.find(View).prop("onSelect")("field1value2");

  const [fieldName, fieldValue, filterType] = params.addFilter.mock.calls[0];
  expect(fieldName).toEqual("field1");
  expect(fieldValue).toEqual("field1value2");
  expect(filterType).toEqual("any");
});

it("will overwrite a filter when a facet value is selected with onChange", () => {
  const wrapper = shallow(<FacetContainer {...params} filterType="any" />);

  wrapper.find(View).prop("onChange")("field1value2");

  const [fieldName, fieldValue, filterType] = params.setFilter.mock.calls[0];
  expect(fieldName).toEqual("field1");
  expect(fieldValue).toEqual("field1value2");
  expect(filterType).toEqual("any");
});

it("will remove a filter when a facet value removed", () => {
  const wrapper = shallow(<FacetContainer {...params} filterType="any" />);

  wrapper.find(View).prop("onRemove")("field1value2");

  const [fieldName, fieldValue, filterType] = params.removeFilter.mock.calls[0];
  expect(fieldName).toEqual("field1");
  expect(fieldValue).toEqual("field1value2");
  expect(filterType).toEqual("any");
});

it("will remove a filter when a facet value removed, defaulting filterType to all", () => {
  const wrapper = shallow(<FacetContainer {...params} />);

  wrapper.find(View).prop("onRemove")("field1value2");

  const [fieldName, fieldValue, filterType] = params.removeFilter.mock.calls[0];
  expect(fieldName).toEqual("field1");
  expect(fieldValue).toEqual("field1value2");
  expect(filterType).toEqual("all");
});

it("passes className through to the view", () => {
  let viewProps;
  const className = "test-class";
  shallow(
    <FacetContainer
      {...params}
      className={className}
      view={props => (viewProps = props)}
    />
  );
  expect(viewProps.className).toEqual(className);
});

it("passes data-foo through to the view", () => {
  let viewProps;
  const data = "bar";
  shallow(
    <FacetContainer
      {...params}
      data-foo={data}
      view={props => (viewProps = props)}
    />
  );
  expect(viewProps["data-foo"]).toEqual(data);
});

describe("search facets", () => {
  let wrapper;
  const field = "field1";
  const fieldData = [
    { count: 20, value: "Virat" },
    { count: 9, value: "LÒpez" },
    { count: 8, value: "bumŗÄh" }
  ];

  function subject(additionalProps = {}, data = fieldData) {
    return shallow(
      <FacetContainer
        {...{
          ...params,
          field,
          facets: {
            field1: [
              {
                field,
                data,
                type: "value"
              }
            ]
          },
          isFilterable: true,
          ...additionalProps
        }}
      />
    );
  }

  beforeAll(() => {
    wrapper = subject();
  });

  it("should have a search input", () => {
    expect(wrapper.find(View).prop("showSearch")).toEqual(true);
  });

  it("should use the field name as a search input placeholder", () => {
    expect(wrapper.find(View).prop("searchPlaceholder")).toBe(
      `Filter ${field}`
    );
  });

  describe("after a search is performed", () => {
    it("should match Facet options with/without accented characters", () => {
      wrapper.find(View).prop("onSearch")("ra");

      const filteredOptions = wrapper.find(View).prop("options");

      expect(filteredOptions.length).toEqual(2);
      expect(filteredOptions.map(opt => opt.value)).toEqual([
        "Virat",
        "bumŗÄh"
      ]);
    });

    it("should ignore case sensitive when matching", () => {
      const data = [
        { count: 20, value: "APPLE" },
        { count: 10, value: "appointment" },
        { count: 9, value: "entertainMEnt" }
      ];
      const wrapper = subject({}, data);

      // action => lowercase
      wrapper.find(View).prop("onSearch")("app");

      const options1 = wrapper.find(View).prop("options");

      expect(options1.length).toEqual(2);
      expect(options1.map(opt => opt.value)).toEqual(["APPLE", "appointment"]);

      // action => uppercase
      wrapper.find(View).prop("onSearch")("MENT");

      const options2 = wrapper.find(View).prop("options");

      expect(options1.length).toEqual(2);
      expect(options2.map(opt => opt.value)).toEqual([
        "appointment",
        "entertainMEnt"
      ]);
    });

    it("should not render Facet options if search value not matched", () => {
      wrapper.find(View).prop("onSearch")("No match");

      expect(wrapper.find(View).prop("options").length).toEqual(0);
    });
  });

  it("should hide the search input", () => {
    const newWrapper = subject({ isFilterable: false });

    expect(newWrapper.find(View).prop("showSearch")).toEqual(false);
  });
});
