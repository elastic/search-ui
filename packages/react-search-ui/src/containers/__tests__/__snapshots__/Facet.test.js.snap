// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  label="Field 1"
  onChange={[Function]}
  onMoreClick={[Function]}
  onRemove={[Function]}
  onSearch={[Function]}
  onSelect={[Function]}
  options={
    Array [
      Object {
        "count": 20,
        "selected": true,
        "value": "field1value1",
      },
      Object {
        "count": 10,
        "selected": false,
        "value": "field1value2",
      },
    ]
  }
  searchPlaceholder="Filter field1"
  showMore={false}
  showSearch={false}
  values={
    Array [
      "field1value1",
    ]
  }
/>
`;
